File: ra-idp/src/authentication/usecases/verify-totp-factor.test.ts
import { describe, expect, it } from 'bun:test'
import { InMemoryMfaFactorRepository } from '../../../adapters/persistence/memory/mfa-factor-repo'
import { generateTotp } from './totp'
import { verifyTotpFactorUseCase } from './verify-totp-factor'

const SECRET = 'GEZDGNBVGY3TQOJQGEZDGNBVGY3TQOJQ'

describe('verifyTotpFactorUseCase', () => {
  it('factor 未登録なら no_factor', async () => {
    const mfaFactorRepo = new InMemoryMfaFactorRepository()
    const result = await verifyTotpFactorUseCase(
      { mfaFactorRepo },
      { sub: 'user-1', code: '000000' },
    )
    expect(result).toEqual({ ok: false, reason: 'no_factor' })
  })

  it('正しいコードを受理し last_used_at を更新する', async () => {
    const mfaFactorRepo = new InMemoryMfaFactorRepository()
    await mfaFactorRepo.save({
      sub: 'user-1',
      type: 'totp',
      secret: SECRET,
      created_at: new Date('2024-01-01T00:00:00Z').toISOString(),
    })
    const now = new Date('2025-06-01T00:00:00Z')
    const code = generateTotp(SECRET, Math.floor(now.getTime() / 1000))

    const result = await verifyTotpFactorUseCase({ mfaFactorRepo }, { sub: 'user-1', code }, now)
    expect(result).toEqual({ ok: true })

    const stored = await mfaFactorRepo.find('user-1', 'totp')
    expect(stored?.last_used_at).toBe(now.toISOString())
  })

  it('誤ったコードは invalid_code', async () => {
    const mfaFactorRepo = new InMemoryMfaFactorRepository()
    await mfaFactorRepo.save({
      sub: 'user-1',
      type: 'totp',
      secret: SECRET,
      created_at: new Date('2024-01-01T00:00:00Z').toISOString(),
    })
    const result = await verifyTotpFactorUseCase(
      { mfaFactorRepo },
      { sub: 'user-1', code: '000000' },
      new Date('2025-06-01T00:00:00Z'),
    )
    expect(result).toEqual({ ok: false, reason: 'invalid_code' })
  })
})

File: ra-idp-go/internal/authentication/usecases/verify_totp_factor_test.go
package usecases

import (
	"context"
	"testing"
	"time"

	"ra-idp-go/internal/infrastructure/persistence/memory"
	"ra-idp-go/internal/spec"
)

func TestVerifyTOTPFactorReturnsNoFactor(t *testing.T) {
	repo := memory.NewMfaFactorRepository()
	result, err := VerifyTOTPFactor(context.Background(), repo, "user-1", "123456", time.Unix(59, 0))
	if err != nil {
		t.Fatalf("verify totp factor: %v", err)
	}
	if result.OK || result.Reason != "no_factor" {
		t.Fatalf("result=%+v, want no_factor", result)
	}
}

func TestVerifyTOTPFactorUpdatesLastUsedAt(t *testing.T) {
	repo := memory.NewMfaFactorRepository()
	secret := rfc6238SHA1SecretBase32
	created := time.Unix(1, 0).UTC()
	now := time.Unix(59, 0).UTC()
	if err := repo.Save(context.Background(), &spec.MfaFactor{
		Sub: "user-1", Type: spec.MfaFactorTOTP, Secret: &secret, CreatedAt: created,
	}); err != nil {
		t.Fatalf("save factor: %v", err)
	}
	code, err := GenerateTOTP(secret, now.Unix())
	if err != nil {
		t.Fatalf("generate totp: %v", err)
	}
	result, err := VerifyTOTPFactor(context.Background(), repo, "user-1", code, now)
	if err != nil {
		t.Fatalf("verify totp factor: %v", err)
	}
	if !result.OK {
		t.Fatalf("result=%+v, want ok", result)
	}
	stored, err := repo.Find(context.Background(), "user-1", spec.MfaFactorTOTP)
	if err != nil {
		t.Fatalf("find factor: %v", err)
	}
	if stored.LastUsedAt == nil || !stored.LastUsedAt.Equal(now) {
		t.Fatalf("last_used_at=%v, want %v", stored.LastUsedAt, now)
	}
}

func TestVerifyTOTPFactorRejectsInvalidCode(t *testing.T) {
	repo := memory.NewMfaFactorRepository()
	secret := rfc6238SHA1SecretBase32
	if err := repo.Save(context.Background(), &spec.MfaFactor{
		Sub: "user-1", Type: spec.MfaFactorTOTP, Secret: &secret, CreatedAt: time.Unix(1, 0).UTC(),
	}); err != nil {
		t.Fatalf("save factor: %v", err)
	}
	result, err := VerifyTOTPFactor(context.Background(), repo, "user-1", "000000", time.Unix(59, 0))
	if err != nil {
		t.Fatalf("verify totp factor: %v", err)
	}
	if result.OK || result.Reason != "invalid_code" {
		t.Fatalf("result=%+v, want invalid_code", result)
	}
}
